system/jlib/jmutex.hpp

```diff
@@ -1002,22 +1002,37 @@ template <typename X>
 class Singleton
 {
 public:
-    template <typename FUNC> X * query(FUNC factory) { return querySingleton(singleton, cs, factory); }
+    template <typename FUNC> X * query(FUNC factory)
+    {
+        if (initialized.load(std::memory_order_acquire))
+            return singleton.load(std::memory_order_acquire); // avoid crit
+        CriticalBlock block(cs);
+        X * value = singleton.load(std::memory_order_acquire); // reload in case another thread got here first
+        if (value) // implies initialized=true also
```

Comment:
should be checking initialised rather than value - in case the factory is expensive, or has side-effects.

Reply:
changed